### PR TITLE
Refactor installer and build to allow arm builds targeted at grub platforms

### PR DESCRIPTION
**Pull request by alexrallen**
Refactors the SONiC Installer to support greater flexibility in building for a given architecture and bootloader. 

Currently the SONiC installer assumes that if a platform is ARM based that it uses the `uboot` bootloader and uses the `grub` bootloader otherwise. This is not a correct assumption to make as ARM is not strictly tied to uboot and x86 is not strictly tied to grub. 

To implement this I introduce the following changes:
* Remove the different arch folders from the `installer/` directory
* Merge the generic components of the ARM and x86 installer into `installer/installer.sh`
* Refactor x86 + grub specific functions into `installer/default_platform.conf` 
* Modify installer to call `default_platform.conf` file and also call `platform/[platform]/patform.conf` file as well to override as needed
* Update references to the installer in the `build_image.sh` script
* Add `TARGET_BOOTLOADER` variable that is by default `uboot` for ARM devices and `grub` for x86 unless overridden in `platform/[platform]/rules.mk`
* Update bootloader logic in `build_debian.sh` to be based on `TARGET_BOOTLOADER` instead of `TARGET_ARCH` and to reference the grub package in a generic manner

This has been tested on a development mellanox platform based on ARM

### CODE REVIEW

File: installer/default_platform.conf [hunk 1]
```diff
@@ -1,119 +1,5 @@
 #!/bin/sh
 
-#  Copyright (C) 2014-2015 Curt Brune <curt@cumulusnetworks.com>
```
Comment (dgsudharsan) [inline installer/default_platform.conf:3]: Why do you want to remove the headers and copyright?

File: installer/default_platform.conf [hunk 2]
```diff
@@ -457,158 +349,72 @@ demo_install_uefi_grub()
...
+        CSTATES=""
+    fi
 
-DEFAULT_GRUB_SERIAL_COMMAND="serial --port=${CONSOLE_PORT} --speed=${CONSOLE_SPEED} --word=8 --parity=no --stop=1"
-DEFAULT_GRUB_CMDLINE_LINUX="console=tty0 console=ttyS${CONSOLE_DEV},${CONSOLE_SPEED}n8 quiet $CSTATES"
-GRUB_SERIAL_COMMAND=${GRUB_SERIAL_COMMAND:-"$DEFAULT_GRUB_SERIAL_COMMAND"}
-GRUB_CMDLINE_LINUX=${GRUB_CMDLINE_LINUX:-"$DEFAULT_GRUB_CMDLINE_LINUX"}
-export GRUB_SERIAL_COMMAND
-export GRUB_CMDLINE_LINUX
+    DEFAULT_GRUB_SERIAL_COMMAND="serial --speed=115200 --word=8 --parity=no --stop=1"
```
Comment (dgsudharsan) [inline installer/default_platform.conf]: Why is this change required? Why are we hardcoding speed here to 115200?

**Review by dgsudharsan — CHANGES_REQUESTED** [review]

File: build_debian.sh [hunk 3]
```diff
@@ -392,11 +392,17 @@ sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get -y in
     systemd-sysv \
     ntp
 
-if [[ $CONFIGURED_ARCH == amd64 ]]; then
+if [[ $TARGET_BOOTLOADER == grub ]]; then
+    if [[ $CONFIGURED_ARCH == amd64 ]]; then
+        GRUB_PKG=grub-pc-bin
+    elif [[ $CONFIGURED_ARCH == arm64 ]]; then
+        GRUB_PKG=grub-efi-arm64-bin
+    fi
```
Comment (vivekrnv) [inline build_debian.sh:400]: What about armhf?

File: build_image.sh [hunk 4]
```diff
@@ -111,7 +111,7 @@ if [ "$IMAGE_TYPE" = "onie" ]; then
     mkdir -p `dirname $OUTPUT_ONIE_IMAGE`
     sudo rm -f $OUTPUT_ONIE_IMAGE
 
-    generate_device_list "./installer/$TARGET_PLATFORM/platforms_asic"
+    generate_device_list "./installer/platforms_asic"
```
Comment (vivekrnv) [inline build_image.sh:114]: I think it's better if you use the output of `extract=1 bash ./sonic-mellanox.bin` to show the diff b/w what you've changed in the contents of the image.

File: files/build_templates/sonic_debian_extension.j2 [hunk 5]
```diff
@@ -294,7 +294,7 @@ sudo dpkg --root=$FILESYSTEM_ROOT -i $debs_path/libnss-radius_*.deb || \
 sudo sed -i -e '/^passwd/s/ radius//' $FILESYSTEM_ROOT/etc/nsswitch.conf
 
 # Install a custom version of kdump-tools  (and its dependencies via 'apt-get -y install -f')
-if [[ $CONFIGURED_ARCH == amd64 ]]; then
+if [[ $TARGET_BOOTLOADER == grub ]]; then
```
Comment (vivekrnv) [inline files/build_templates/sonic_debian_extension.j2:297]: I don't follow how is kdump-tools deb related to the type of bootloader the device uses.

File: installer/install.sh [hunk 6]
```diff
@@ -1,7 +1,16 @@
 #!/bin/sh
 
-#  Copyright (C) Marvell Inc
+#  Copyright (C) 2020      Marvell Inc
+#  Copyright (C) 2014-2015 Curt Brune <curt@cumulusnetworks.com>
```
Comment (vivekrnv) [inline installer/install.sh:4]: Do you think it's better to move create_partition & mount_partition methods in the default_platform.conf into a different file which is specific to x86?

Also, can you briefly summarize the refactoring you did in the installer folder in the PR description?

File: build_debian.sh (hunk 3 above)
Comment (alexrallen) [reply build_debian.sh:400]: armhf is irrelevant here because these lines are concerning the grub package to install from APT. 

There are two types of platforms which now use grub in SONiC:
1. All amd64 platforms
2. Mellanox arm64 platforms

No armhf platforms support grub. As that would be packages and code I would be unable to test I do not want to include it, that instead would be the responsibility of someone trying to add support for a platform which supports grub on armhf.

File: build_image.sh (hunk 4 above)
Comment (alexrallen) [reply build_image.sh:114]: Can you please clarify your suggestion here? 

The difference you have highlighted is simply the result of me relocating the `platforms_asic` file because there are now no longer arch-specific installer directories, there is just a common installer.

File: files/build_templates/sonic_debian_extension.j2 (hunk 5 above)
Comment (alexrallen) [reply files/build_templates/sonic_debian_extension.j2:297]: kdump currently does not support uboot. Please see the following commit taken from the git-blame of this line.

https://github.com/Azure/sonic-buildimage/pull/3962

File: installer/install.sh (hunk 6 above)
Comment (alexrallen) [reply installer/install.sh:4]: Currently SONiC makes an assumption that all platforms that are amd64: 

1. Support grub
2. Are architect-ed in such a way that these mount / unmount functions will work 

Revoking this assumption would mean copying these functions to the platform-specific `platform.conf` code of many other platforms which would involve a fair amount of repeated code and also involve modifying many non-mellanox platform files. 

Generally we try to avoid modifying platform-specific code outside of Mellanox as that is heavily scrutinized in the community. 

With this we include the create_partition and mount_partition in the `default_platform.conf` file as that configuration is currently considered the "default" platform in SONiC. Any other platforms can override any one or more of these functions as needed.

File: installer/default_platform.conf (hunk 1 above)
Comment (alexrallen) [reply installer/default_platform.conf:3]: Wasn't sure how to handle this, all the headers and copyright are now merged and located in installer.sh would you like me to move that to this file as well?

File: installer/default_platform.conf (hunk 2 above)
Comment (alexrallen) [reply installer/default_platform.conf]: This is an OrinX change I erroneously included in this branch. I will fix.

File: installer/default_platform.conf (hunk 1 above)
Comment (alexrallen) [reply installer/default_platform.conf:3]: You can see the vast majority of this file has been changed... unsure how to appropriately preserve copyright here.

**Review by dgsudharsan — APPROVED** [review]
